instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+        """
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+
+        return cls._instance
+
+    def __init__(
+        self,
+        methods: List[str] = None,
+        recurse_attribs: Dict[Type[BaseEstimator], List[str]] = None,
+        recurse_namedtuple_attribs: Dict[
+            Type[BaseEstimator], List[str]
+        ] = None,
+        spanner: Callable = implement_spans,
+        packages: List[str] = None,
+        exclude_classes: List[Type] = None,
+    ):
+        self.methods = methods or DEFAULT_METHODS
+        self.recurse_attribs = recurse_attribs or DEFAULT_ATTRIBS
+        self.recurse_namedtuple_attribs = (
+            recurse_namedtuple_attribs or DEFAULT_NAMEDTUPLE_ATTRIBS
+        )
+        self.spanner = spanner
+        self.packages = packages or DEFAULT_PACKAGES
+        if exclude_classes is None:
+            self.exclude_classes = tuple(DEFAULT_EXCLUDE_CLASSES)
+        else:
+            self.exclude_classes = tuple(exclude_classes)
+
+    def _instrument(self, **kwargs):
+        """Instrument the library, and any additional specified on init."""
+        klasses = get_base_estimators(packages=self.packages)
+        for _, klass in klasses.items():
+            if issubclass(klass, self.exclude_classes):
+                logger.debug("Not instrumenting (excluded): %s", str(klass))
+            else:
+                logger.debug("Instrumenting: %s", str(klass))
+                for method_name in self.methods:
+                    if hasattr(klass, method_name):
+                        self._instrument_class_method(
+                            estimator=klass, method_name=method_name
+                        )
+
+    def _uninstrument(self, **kwargs):
+        """Uninstrument the library"""
+        klasses = get_base_estimators(packages=self.packages)
+        for _, klass in klasses.items():
+            logger.debug("Uninstrumenting: %s", str(klass))
+            for method_name in self.methods:
+                if hasattr(klass, method_name):
+                    self._uninstrument_class_method(
+                        estimator=klass, method_name=method_name
+                    )
+
+    def instrument_estimator(
+        self, estimator: BaseEstimator, attributes: Attributes = None
+    ):
+        """Instrument a fitted estimator and its hierarchy where configured.
+
+        Args:
+            estimator (sklearn.base.BaseEstimator): A fitted ``sklearn``
+              estimator, typically a ``Pipeline`` instance.
+            attributes (dict): Attributes to attach to the spans.
+        """
+        if isinstance(estimator, self.exclude_classes):
+            logger.debug(
+                "Not instrumenting (excluded): %s",
+                estimator.__class__.__name__,
+            )
+            return
+
+        if isinstance(
+            estimator, tuple(self.recurse_namedtuple_attribs.keys())
```

Comment:
A `Pipeline` is a tuple?

Reply:
We're not checking if the estimator is a tuple. Tuples are the means by which we use `isinstance` to check if an object is one of multiple types. In this case we are checking if the estimator is one of the classes defined in the `recurse_namedtuple_attribs` set of keys, that way we know to proceed with continuing iteration through the corresponding values.

Reply:
Whoops my mistake, read that wrong.